clap_generate/src/shell.rs

```diff
@@ -19,36 +21,160 @@ pub enum Shell {
 
 impl Shell {
     /// A list of supported shells in `[&'static str]` form.
-    pub fn variants() -> [&'static str; 5] {
-        ["bash", "elvish", "fish", "powershell", "zsh"]
+    pub fn variants() -> &'static [&'static str] {
+        &VARIANTS
     }
 }
 
 impl Display for Shell {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match *self {
-            Shell::Bash => write!(f, "bash"),
-            Shell::Elvish => write!(f, "elvish"),
-            Shell::Fish => write!(f, "fish"),
-            Shell::PowerShell => write!(f, "powershell"),
-            Shell::Zsh => write!(f, "zsh"),
-        }
+        self.as_arg().expect("no values are skipped").fmt(f)
     }
 }
 
 impl FromStr for Shell {
     type Err = String;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s.to_ascii_lowercase().as_str() {
-            "bash" => Ok(Shell::Bash),
-            "elvish" => Ok(Shell::Elvish),
-            "fish" => Ok(Shell::Fish),
-            "powershell" => Ok(Shell::PowerShell),
-            "zsh" => Ok(Shell::Zsh),
-            _ => Err(String::from(
-                "[valid values: bash, elvish, fish, powershell, zsh]",
-            )),
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                current_str.eq_ignore_ascii_case(s).then(|| *current_value)
+            })
+            .next()
+            .ok_or_else(|| String::from(VALID_VALUES))
+    }
+}
+
+// Hand-rolled so it can work even when `derive` feature is disabled
+impl ArgEnum for Shell {
+    const VARIANTS: &'static [&'static str] = &VARIANTS;
+
+    /// Parse an argument into `Self`.
+    fn from_str(input: &str, case_insensitive: bool) -> Result<Self, String> {
+        let eq = if case_insensitive {
+            str::eq_ignore_ascii_case
+        } else {
+            PartialEq::eq
+        };
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                eq(current_str, input).then(|| *current_value)
+            })
+            .next()
+            .ok_or_else(|| String::from(VALID_VALUES))
+    }
+
+    fn as_arg(&self) -> Option<&'static str> {
+        MAPPINGS
+            .iter()
+            .filter_map(|(current_str, current_value)| {
+                (self == current_value).then(|| *current_str)
+            })
+            .next()
+    }
+}
+
+impl crate::Generator for Shell {
+    fn file_name(&self, name: &str) -> String {
+        match self {
+            Shell::Bash => crate::generators::Bash.file_name(name),
+            Shell::Elvish => crate::generators::Elvish.file_name(name),
+            Shell::Fish => crate::generators::Fish.file_name(name),
+            Shell::PowerShell => crate::generators::PowerShell.file_name(name),
+            Shell::Zsh => crate::generators::Zsh.file_name(name),
```

Comment:
Can we please import these at the top of the file?

Reply:
I think in this case, I disagree.  While I personally prefer `use`ing only a minimum set of non-traits because I find it makes the code clearer and less likely to produce name collisions, here I think its important to make intent clear with `Shell::Zsh` vs `crate::generators::Zsh` (rather than `Shell::Zsh` vs `Zsh`).